Use template in a function with more than 15 entries print warnings

### Describe the bug

I have a function with a variable defined by a template with at least 2 possible types and more than 16 arguments. When I try to use Pyccel on such a function, I get warning messages about some integer variable being too large. To be exact, I get 2 messages if I have 16 arguments, 4 messages if I have 17 arguments, etc... 

These warnings appear only if the function has an argument defined by a template with at least 2 types. If a unique type is defined in the template, no warning messages appear.

### To Reproduce

```python
from pyccel.decorators import template

@template(name='T', types=['float[:,:,:,:,:]','complex[:,:,:,:,:]'])
def func(M: 'T', n1: "int64", n2: "int64", n3: "int64", n4: "int64", n5: "int64",
                 p1: "int64", p2: "int64", p3: "int64", p4: "int64", p5: "int64",
                 a1: "int64", a2: "int64", a3: "int64", a4: "int64", a5: "int64"):

    M[n1, n2, n3, n4, n5] = M[p1, p2, p3, p4, p5] + a1 + a2 + a3 + a4 + a5
    return

```
I tried to run the function in the terminal with the command line :
"pyccel file_name.py"

### Error details

```
~/Code/psydac/psydac/api/ast$ pyccel kernel_pyccel.py
/home/cato/venv/psydac/lib/python3.8/site-packages/pyccel/codegen/compiling/compilers.py:431: UserWarning: /home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c: In function ‘type_check’:
/home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c:132:1: warning: integer constant is so large that it is unsigned
  132 |         check += INT64_C(11529215046068469760);
      | ^       ~~~~~~~~~~~~~
/home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c: In function ‘add_2d_wrapper’:
/home/cato/Code/psydac/psydac/api/ast/__pyccel__/kernel_pyccel_wrapper.c:391:1: warning: integer constant is so large that it is unsigned
  391 |     else if (check == INT64_C(11682937913349382690))
      | ^   ~~~~~~~~~~~~~~~~~

  warnings.warn(UserWarning(err))

```

### Expected behavior
It should not print warning message.

### Language
The Python code is translated to Fortran.

--- a/pyccel/ast/cwrapper.py
+++ b/pyccel/ast/cwrapper.py
@@ -531,23 +531,3 @@ def scalar_object_check(py_object, c_object):
                     results   = [Variable(dtype=NativeBool(), name = 'r')])
 
     return FunctionCall(check_func, [py_object])
-
-# This registry is used for interface management,
-# mapping each data type to a given flag
-# Those flags are used in a bitset #TODO
-flags_registry = {
-    (NativeInteger(), 4)       : 1,
-    (NativeInteger(), 8)       : 2,
-    (NativeInteger(), 2)       : 3,
-    (NativeInteger(), 1)       : 4,
-    (NativeInteger(), -1)      : 5,
-    (NativeFloat(), 8)         : 6,
-    (NativeFloat(), 4)         : 7,
-    (NativeFloat(), -1)        : 8,
-    (NativeComplex(), 4)       : 9,
-    (NativeComplex(), 8)       : 10,
-    (NativeComplex(), -1)      : 11,
-    (NativeBool(), 4)          : 12,
-    (NativeBool(), -1)         : 12,
-    (NativeString(), 0)        : 13
-}

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -23,7 +23,7 @@ from pyccel.ast.cwrapper    import PyArg_ParseTupleNode, PyBuildValueNode
 from pyccel.ast.cwrapper    import PyArgKeywords
 from pyccel.ast.cwrapper    import Py_None, Py_DECREF
 from pyccel.ast.cwrapper    import generate_datatype_error, PyErr_SetString
-from pyccel.ast.cwrapper    import scalar_object_check, flags_registry
+from pyccel.ast.cwrapper    import scalar_object_check
 from pyccel.ast.cwrapper    import PyccelPyArrayObject, PyccelPyObject
 from pyccel.ast.cwrapper    import C_to_Python, Python_to_C
 from pyccel.ast.cwrapper    import PyModule_AddObject
@@ -39,7 +39,7 @@ from pyccel.ast.numpy_wrapper   import pyarray_to_ndarray
 from pyccel.ast.numpy_wrapper   import array_get_data, array_get_dim
 
 from pyccel.ast.operators import PyccelEq, PyccelNot, PyccelOr, PyccelAssociativeParenthesis
-from pyccel.ast.operators import PyccelIsNot, PyccelLt, PyccelUnarySub
+from pyccel.ast.operators import PyccelIsNot, PyccelLt, PyccelUnarySub, PyccelNe
 
 from pyccel.ast.variable  import Variable, DottedVariable
 
@@ -971,6 +971,17 @@ class CWrapperCodePrinter(CCodePrinter):
         # collect parse arg
         parse_args = [self.get_PyArgParseType(a.var) for a in funcs[0].arguments]
 
+        # Determine flags which indicate argument type
+        argument_type_flags = {func:[] for func in funcs}
+        step = 1
+        for i, p_arg in enumerate(parse_args):
+            interface_args = [func.arguments[i].var for func in funcs]
+            interface_types = [(a.dtype, a.precision) for a in interface_args]
+            possible_types = list(dict.fromkeys(interface_types)) # Remove duplicates but preserve order
+            for func, t in zip(funcs, interface_types):
+                argument_type_flags[func].append(possible_types.index(t)*step)
+            step *= len(possible_types)
+
         # Managing the body of wrapper
         for func in funcs :
             mini_wrapper_func_body = []
@@ -990,10 +1001,9 @@ class CWrapperCodePrinter(CCodePrinter):
                 mini_scope.insert_variable(a)
             for r in func.results:
                 mini_scope.insert_variable(r)
-            flags = 0
 
             # Loop for all args in every functions and create the corresponding condition and body
-            for p_arg, (f_var, f_arg) in zip(parse_args, local_arg_vars.items()):
+            for idx, (p_arg, (f_var, f_arg)) in enumerate(zip(parse_args, local_arg_vars.items())):
                 collect_var  = self.get_PyArgParseType(f_var)
                 body, tmp_variable = self._body_management(f_var, p_arg, f_arg.value)
 
@@ -1010,9 +1020,11 @@ class CWrapperCodePrinter(CCodePrinter):
                 # Get Bind/C arguments
                 static_func_args.extend(self.get_static_args(f_var))
 
-                flag_value = flags_registry[(f_var.dtype, f_var.precision)]
-                flags = (flags << 4) + flag_value  # shift by 4 to the left
-                types_dict[f_var].add((f_var, check, flag_value)) # collect variable type for each arguments
+                # Save flag to types dict for interface recognition
+                flag_value = argument_type_flags[func][idx]
+                if flag_value >= len(types_dict[f_var]):
+                    types_dict[f_var].add((f_var, check, flag_value)) # collect variable type for each arguments
+
                 mini_wrapper_func_body += body
 
             # create the corresponding function call
@@ -1048,7 +1060,7 @@ class CWrapperCodePrinter(CCodePrinter):
             funcs_def.append(mini_wrapper_func_def)
 
             # append check condition to the functioncall
-            body_tmp.append(IfSection(PyccelEq(check_var, LiteralInteger(flags)), [AliasAssign(wrapper_results[0],
+            body_tmp.append(IfSection(PyccelEq(check_var, LiteralInteger(sum(argument_type_flags[func]))), [AliasAssign(wrapper_results[0],
                     FunctionCall(mini_wrapper_func_def, parse_args))]))
 
         # Errors / Types management
@@ -1057,7 +1069,7 @@ class CWrapperCodePrinter(CCodePrinter):
         funcs_def.append(check_func_def)
 
         # Create the wrapper body with collected informations
-        body_tmp = [IfSection(PyccelNot(check_var), [Return([Nil()])])] + body_tmp
+        body_tmp = [IfSection(PyccelNe(check_var, PyccelUnarySub(LiteralInteger(1))), [Return([Nil()])])] + body_tmp
         body_tmp.append(IfSection(LiteralTrue(),
             [PyErr_SetString('PyExc_TypeError', '"This combination of arguments is not valid"'),
             Return([Nil()])]))
@@ -1090,31 +1102,57 @@ class CWrapperCodePrinter(CCodePrinter):
         return sep + '\n'.join(CCodePrinter._print_FunctionDef(self, f) for f in funcs_def)
 
     def _create_wrapper_check(self, check_var, parse_args, types_dict, func_name):
+        """
+        Generate the function which determines the relevant interface.
+
+        Parameters
+        ----------
+        check_var : Variable
+            The integer variable which will be returned by the function.
+
+        parse_args : list of Variable
+            The arguments passed to the function in the Python code.
+
+        types_dict : dictionary
+            A dictionary linking the parsed arguments to the possible types.
+            The values are tuples containing the expected argument, the type check, and the flag.
+
+        func_name : str
+            The name of the interface.
+
+        Returns
+        -------
+        FunctionDef
+            The function which determines the Interface key
+        """
         check_func_body = []
-        flags = (len(types_dict) - 1) * 4
+        step = 1
         for arg in types_dict:
             var_name = ""
             body = []
-            types = []
             arg_type_check_list = list(types_dict[arg])
-            arg_type_check_list.sort(key= lambda x : x[0].precision)
-            for elem in arg_type_check_list:
-                var_name = elem[0].name
-                value = elem[2] << flags
-                body.append(IfSection(elem[1], [AugAssign(check_var, '+' ,value)]))
-                types.append(elem[0])
-            flags -= 4
+            types = [elem[0] for elem in arg_type_check_list]
             error = ' or '.join(['{} {}'.format(str(v.precision * 8)+' bit' if v.precision != -1 else 'native',
                                                     str_dtype(v.dtype))
                             if not isinstance(v.dtype, NativeBool)
                             else  str_dtype(v.dtype) for v in types])
-            body.append(IfSection(LiteralTrue(), [PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(var_name, error)), Return([LiteralInteger(0)])]))
+            if len(arg_type_check_list) == 1:
+                elem = arg_type_check_list[0]
+                var_name = elem[0].name
+                assert elem[2] == 0
+                body.append(IfSection(PyccelNot(elem[1]), [PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(var_name, error)), Return([PyccelUnarySub(LiteralInteger(1))])]))
+            else:
+                arg_type_check_list.sort(key= lambda x : x[2])
+                for elem in arg_type_check_list:
+                    var_name = elem[0].name
+                    body.append(IfSection(elem[1], [AugAssign(check_var, '+' ,elem[2])]))
+                body.append(IfSection(LiteralTrue(), [PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(var_name, error)), Return([PyccelUnarySub(LiteralInteger(1))])]))
             check_func_body += [If(*body)]
 
         check_func_body = [Assign(check_var, LiteralInteger(0))] + check_func_body
         check_func_body.append(Return([check_var]))
         # Creating check function definition
-        check_func_name = self.scope.parent_scope.get_new_name('type_check')
+        check_func_name = self.scope.parent_scope.get_new_name(f'type_check_{func_name}')
         check_func_def = FunctionDef(name = check_func_name,
             arguments = parse_args,
             results = [check_var],

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -1105,6 +1105,12 @@ class CWrapperCodePrinter(CCodePrinter):
         """
         Generate the function which determines the relevant interface.
 
+        Creates a function which tests each of the arguments passed to the function.
+        For each of the arguments it checks if it has one of the expected types. If
+        this is not the case, then an error is raised, otherwise the received type is
+        saved into a byte flag. This flag is then used to determine exactly which
+        function in the interface is being called.
+
         Parameters
         ----------
         check_var : Variable
@@ -1123,7 +1129,7 @@ class CWrapperCodePrinter(CCodePrinter):
         Returns
         -------
         FunctionDef
-            The function which determines the Interface key
+            The function which determines the Interface key.
         """
         check_func_body = []
         step = 1

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -1185,7 +1185,6 @@ class CWrapperCodePrinter(CCodePrinter):
         """
         Generate the function which determines the relevant interface.
 
-
         Creates a FunctionDef which tests each of the arguments passed to the Interface.
         For each of the arguments it checks if it has one of the expected types.
         This function is necessary when wrapping an Interface in order to determine which

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -1091,7 +1091,7 @@ class CWrapperCodePrinter(CCodePrinter):
 
         # Errors / Types management
         # Creating check_type function
-        check_func_def = self._create_wrapper_check(check_var, parse_args, types_dict)
+        check_func_def = self._create_wrapper_check(check_var, parse_args, types_dict, funcs[0].name)
         funcs_def.append(check_func_def)
 
         # Create the wrapper body with collected informations

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -39,7 +39,7 @@ from pyccel.ast.numpy_wrapper   import pyarray_to_ndarray
 from pyccel.ast.numpy_wrapper   import array_get_data, array_get_dim
 
 from pyccel.ast.operators import PyccelEq, PyccelNot, PyccelOr, PyccelAssociativeParenthesis
-from pyccel.ast.operators import PyccelIsNot, PyccelLt, PyccelUnarySub, PyccelNe
+from pyccel.ast.operators import PyccelIsNot, PyccelLt, PyccelUnarySub
 
 from pyccel.ast.variable  import Variable, DottedVariable
 
@@ -1069,7 +1069,7 @@ class CWrapperCodePrinter(CCodePrinter):
         funcs_def.append(check_func_def)
 
         # Create the wrapper body with collected informations
-        body_tmp = [IfSection(PyccelNe(check_var, PyccelUnarySub(LiteralInteger(1))), [Return([Nil()])])] + body_tmp
+        body_tmp = [IfSection(PyccelEq(check_var, PyccelUnarySub(LiteralInteger(1))), [Return([Nil()])])] + body_tmp
         body_tmp.append(IfSection(LiteralTrue(),
             [PyErr_SetString('PyExc_TypeError', '"This combination of arguments is not valid"'),
             Return([Nil()])]))

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -1154,8 +1154,8 @@ class CWrapperCodePrinter(CCodePrinter):
         Creates a function which tests each of the arguments passed to the function.
         For each of the arguments it checks if it has one of the expected types. If
         this is not the case, then an error is raised, otherwise the received type is
-        saved into a byte flag. This flag is then used to determine exactly which
-        function in the interface is being called.
+        used to increment a type flag. This flag is then used to determine exactly which
+        function in the interface is being called (see _determine_interface_flags for more detail).
 
         Parameters
         ----------

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -1132,27 +1132,26 @@ class CWrapperCodePrinter(CCodePrinter):
             The function which determines the Interface key.
         """
         check_func_body = []
-        step = 1
         for arg in types_dict:
             var_name = ""
             body = []
             arg_type_check_list = list(types_dict[arg])
             types = [elem[0] for elem in arg_type_check_list]
-            error = ' or '.join(['{} {}'.format(str(v.precision * 8)+' bit' if v.precision != -1 else 'native',
-                                                    str_dtype(v.dtype))
-                            if not isinstance(v.dtype, NativeBool)
-                            else  str_dtype(v.dtype) for v in types])
+            precisions = [f'{v.precision * 8} bit' if v.precision != -1 else 'native' for v in types]
+            type_strs = [str_dtype(v.dtype) for v in types]
+            error = ' or '.join([f'{p} {t}' if not isinstance(v.dtype, NativeBool) else t
+                        for p,t,v in zip(precisions, type_strs, types)])
             if len(arg_type_check_list) == 1:
                 elem = arg_type_check_list[0]
                 var_name = elem[0].name
                 assert elem[2] == 0
-                body.append(IfSection(PyccelNot(elem[1]), [PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(var_name, error)), Return([PyccelUnarySub(LiteralInteger(1))])]))
+                body.append(IfSection(PyccelNot(elem[1]), [PyErr_SetString('PyExc_TypeError', f'"{var_name} must be {error}"'), Return([PyccelUnarySub(LiteralInteger(1))])]))
             else:
                 arg_type_check_list.sort(key= lambda x : x[2])
                 for elem in arg_type_check_list:
                     var_name = elem[0].name
                     body.append(IfSection(elem[1], [AugAssign(check_var, '+' ,elem[2])]))
-                body.append(IfSection(LiteralTrue(), [PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(var_name, error)), Return([PyccelUnarySub(LiteralInteger(1))])]))
+                body.append(IfSection(LiteralTrue(), [PyErr_SetString('PyExc_TypeError', f'"{var_name} must be {error}"'), Return([PyccelUnarySub(LiteralInteger(1))])]))
             check_func_body += [If(*body)]
 
         check_func_body = [Assign(check_var, LiteralInteger(0))] + check_func_body

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -972,15 +972,7 @@ class CWrapperCodePrinter(CCodePrinter):
         parse_args = [self.get_PyArgParseType(a.var) for a in funcs[0].arguments]
 
         # Determine flags which indicate argument type
-        argument_type_flags = {func:[] for func in funcs}
-        step = 1
-        for i, p_arg in enumerate(parse_args):
-            interface_args = [func.arguments[i].var for func in funcs]
-            interface_types = [(a.dtype, a.precision) for a in interface_args]
-            possible_types = list(dict.fromkeys(interface_types)) # Remove duplicates but preserve order
-            for func, t in zip(funcs, interface_types):
-                argument_type_flags[func].append(possible_types.index(t)*step)
-            step *= len(possible_types)
+        argument_type_flags = self._determine_interface_flags(funcs, parse_args)
 
         # Managing the body of wrapper
         for func in funcs :
@@ -1100,6 +1092,60 @@ class CWrapperCodePrinter(CCodePrinter):
         self.exit_scope()
 
         return sep + '\n'.join(CCodePrinter._print_FunctionDef(self, f) for f in funcs_def)
+
+    def _determine_interface_flags(self, funcs):
+        """
+        Determine the flags which allow correct function to be identified from the interface.
+
+        Each function must be identifiable by a different integer value. This value is known
+        as a flag. Different parts of the flag to indicate the types of different arguments.
+        Take for example the following function:
+        ```python
+        @types('int', 'int')
+        @types('float', 'float')
+        def f(a, b):
+            pass
+        ```
+        The values 0 (int) and 1 (float) would indicate the type of the argument a. In order
+        to preserve this information the values which indicate the type of the argument b
+        must only change the part of the flag which does not contain this information. In other
+        words `flag % n_types_a = flag_a`. Therefore the values 0 (int) and 2(float) indicate
+        the type of the argument b.
+        We then finally have the following four options:
+          1. 0 = 0 + 0 => (int,int)
+          2. 1 = 1 + 0 => (float,int)
+          3. 2 = 0 + 2 => (int, float)
+          4. 3 = 1 + 2 => (float, float)
+
+        of which only the first and last flags indicate acceptable arguments.
+
+        The function returns a dictionary whose keys are the functions and whose values are
+        a list of the flags which would indicate the correct types.
+        In the above example we would return `{func_0 : [0,0], func_1 : [1,2]}`.
+
+        Parameters
+        ----------
+        funcs : list of FunctionDefs
+            The functions in the Interface.
+
+        Returns
+        -------
+        dict
+            A dictionary whose keys are the functions and whose values are a list of integers
+            which should be used to increment the byte flag if the argument at the same index
+            has the type expected by the function.
+        """
+        argument_type_flags = {func:[] for func in funcs}
+        nargs = len(funcs[0].arguments)
+        step = 1
+        for i in range(nargs):
+            interface_args = [func.arguments[i].var for func in funcs]
+            interface_types = [(a.dtype, a.precision) for a in interface_args]
+            possible_types = list(dict.fromkeys(interface_types)) # Remove duplicates but preserve order
+            for func, t in zip(funcs, interface_types):
+                argument_type_flags[func].append(possible_types.index(t)*step)
+            step *= len(possible_types)
+        return argument_type_flags
 
     def _create_wrapper_check(self, check_var, parse_args, types_dict, func_name):
         """

--- a/pyccel/codegen/printing/cwrappercode.py
+++ b/pyccel/codegen/printing/cwrappercode.py
@@ -1006,7 +1006,7 @@ class CWrapperCodePrinter(CCodePrinter):
         parse_args = [self.get_PyArgParseType(a.var) for a in funcs[0].arguments]
 
         # Determine flags which indicate argument type
-        argument_type_flags = self._determine_interface_flags(funcs, parse_args)
+        argument_type_flags = self._determine_interface_flags(funcs)
 
         # Managing the body of wrapper
         for func in funcs :

--- a/pyccel/version.py
+++ b/pyccel/version.py
@@ -1,4 +1,4 @@
 """
 Module specifying the current version string for pyccel
 """
-__version__ = "1.7.2"
+__version__ = "1.7.3"

